src/server/frontend_wayland/wl_client.h

```diff
@@ -53,6 +57,7 @@ class WlClient : public wayland::LifetimeTracker
         std::function<void(WlClient&)>&& client_created_callback);
 
     static auto from(wl_client* client) -> WlClient*;
+    static auto from_or_throw(wl_client* client) -> WlClient&;
```

Comment:
The naming and semantics of this seem a bit odd. I'd expect the name to reflect the contract and an exception being thrown to indicate the contract cannot be met.

There are some exceptions to this rule (`std::rethrow()` for example), but I don't see the motivation here.

Looking at the implementation, we throw a `logic_error` which also suggests that an exception indicates a precondition failure.

Looking at the use, we either take the address immediately, or simply use a member function. That is, the requirement is is simply a stronger version of the `WlClient::from()` post-condition.

In fact, why can't we just make "result is not `nullptr`" a post-condition of `WlClient::from()`? AFAICS it would need a `logic_error` (or worse) for this to return `nullptr`.